### ISSUE TITLE
DataLoader.load/load_many include any exceptions types in return value

### ISSUE THREAD

**Issue by huonw** [strawberry-graphql/strawberry#1736]
Thanks for Strawberry!

The `DataLoader` class is neat. It seems like the expectation is that the list of loaded values returned from the `load_fn` includes any failures as `Exception`s directly, and they're propagated through the futures, to be re-thrown when `.load()` is awaited:

https://github.com/strawberry-graphql/strawberry/blob/67e6d749fe5652a7ec5788097aa8962770edf34d/strawberry/dataloader.py#L141-L142

For a fully typed code base, this means the load function return value needs to specify the exceptions, like `-> list[Union[MyRealType, Exception]]`. Currently the `load_fn` parameter is typed as:

https://github.com/strawberry-graphql/strawberry/blob/67e6d749fe5652a7ec5788097aa8962770edf34d/strawberry/dataloader.py#L52

so specifying exceptions in the return value like the above sets `T = Union[MyRealType, Exception]`, and this flows through to the return value of the `load` method. However, it's impossible for `load` to ever return an `Exception`, and so code calling `.load` is forced to do type checks/asserts/casts when it tries to do real things with `MyRealType` (e.g. access properties).

It'd be nice if the types were more accurate, so that `load` doesn't need to return the exceptions. One option might be to have `load_fn` have type something like `load_fn: Callable[[List[K]], Awaitable[List[Union[T, Exception]]]],`. The `List` may need to change to `Sequence` to be backwards compatible, because `List`'s type parameter is invariant.

### Code example

`pyproject.toml`:
```toml
[tool.poetry]
name = "strawberry-testing"
version = "0.1.0"
description = ""
authors = []

[tool.poetry.dependencies]
python = "^3.9"
strawberry-graphql = "^0.103.1"
mypy = "^0.941"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

<details><summary>Click for <code>poetry.lock</code></summary>

```toml
[[package]]
name = "backports.cached-property"
version = "1.0.1"
description = "cached_property() - computed once per instance, cached as attribute"
category = "main"
optional = false
python-versions = ">=3.6.0"

[[package]]
name = "click"
version = "8.0.4"
description = "Composable command line interface toolkit"
category = "main"
optional = false
python-versions = ">=3.6"

[package.dependencies]
colorama = {version = "*", markers = "platform_system == \"Windows\""}

[[package]]
name = "colorama"
version = "0.4.4"
description = "Cross-platform colored terminal text."
category = "main"
optional = false
python-versions = ">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*"

[[package]]
name = "graphql-core"
version = "3.2.0"
description = "GraphQL implementation for Python, a port of GraphQL.js, the JavaScript reference implementation for GraphQL."
category = "main"
optional = false
python-versions = ">=3.6,<4"

[[package]]
name = "mypy"
version = "0.941"
description = "Optional static typing for Python"
category = "main"
optional = false
python-versions = ">=3.6"

[package.dependencies]
mypy-extensions = ">=0.4.3"
tomli = ">=1.1.0"
typing-extensions = ">=3.10"

[package.extras]
dmypy = ["psutil (>=4.0)"]
python2 = ["typed-ast (>=1.4.0,<2)"]
reports = ["lxml"]

[[package]]
name = "mypy-extensions"
version = "0.4.3"
description = "Experimental type system extensions for programs checked with the mypy typechecker."
category = "main"
optional = false
python-versions = "*"

[[package]]
name = "pygments"
version = "2.11.2"
description = "Pygments is a syntax highlighting package written in Python."
category = "main"
optional = false
python-versions = ">=3.5"

[[package]]
name = "python-dateutil"
version = "2.8.2"
description = "Extensions to the standard Python datetime module"
category = "main"
optional = false
python-versions = "!=3.0.*,!=3.1.*,!=3.2.*,>=2.7"

[package.dependencies]
six = ">=1.5"

[[package]]
name = "python-multipart"
version = "0.0.5"
description = "A streaming multipart parser for Python"
category = "main"
optional = false
python-versions = "*"

[package.dependencies]
six = ">=1.4.0"

[[package]]
name = "six"
version = "1.16.0"
description = "Python 2 and 3 compatibility utilities"
category = "main"
optional = false
python-versions = ">=2.7, !=3.0.*, !=3.1.*, !=3.2.*"

[[package]]
name = "strawberry-graphql"
version = "0.103.1"
description = "A library for creating GraphQL APIs"
category = "main"
optional = false
python-versions = ">=3.7,<4.0"

[package.dependencies]
"backports.cached-property" = ">=1.0.1,<2.0.0"
click = ">=7.0,<9.0"
graphql-core = ">=3.2.0,<3.3.0"
pygments = ">=2.3,<3.0"
python-dateutil = ">=2.7.0,<3.0.0"
python-multipart = ">=0.0.5,<0.0.6"
typing_extensions = ">=3.7.4,<5.0.0"

[package.extras]
asgi = ["starlette (>=0.13.6,<0.17.0)"]
debug-server = ["starlette (>=0.13.6,<0.17.0)", "uvicorn (>=0.11.6,<0.18.0)"]
django = ["Django (>=2.2)", "asgiref (>=3.2,<4.0)"]
flask = ["flask (>=1.1)"]
opentelemetry = ["opentelemetry-api (<2)", "opentelemetry-sdk (<2)"]
chalice = ["chalice (>=1.22,<2.0)"]
pydantic = ["pydantic (<2)"]
sanic = ["sanic (>=20.12.2,<22.0.0)"]
aiohttp = ["aiohttp (>=3.7.4.post0,<4.0.0)"]
fastapi = ["fastapi (>=0.65.2)"]

[[package]]
name = "tomli"
version = "2.0.1"
description = "A lil' TOML parser"
category = "main"
optional = false
python-versions = ">=3.7"

[[package]]
name = "typing-extensions"
version = "4.1.1"
description = "Backported and Experimental Type Hints for Python 3.6+"
category = "main"
optional = false
python-versions = ">=3.6"

[metadata]
lock-version = "1.1"
python-versions = "^3.9"
content-hash = "01510ad2621d0db4339d8c0626ba91128dabdbc28fdbca4a76a324747b906a9c"

[metadata.files]
"backports.cached-property" = [
    {file = "backports.cached-property-1.0.1.tar.gz", hash = "sha256:1a5ef1e750f8bc7d0204c807aae8e0f450c655be0cf4b30407a35fd4bb27186c"},
    {file = "backports.cached_property-1.0.1-py3-none-any.whl", hash = "sha256:687b5fe14be40aadcf547cae91337a1fdb84026046a39370274e54d3fe4fb4f9"},
]
click = [
    {file = "click-8.0.4-py3-none-any.whl", hash = "sha256:6a7a62563bbfabfda3a38f3023a1db4a35978c0abd76f6c9605ecd6554d6d9b1"},
    {file = "click-8.0.4.tar.gz", hash = "sha256:8458d7b1287c5fb128c90e23381cf99dcde74beaf6c7ff6384ce84d6fe090adb"},
]
colorama = [
    {file = "colorama-0.4.4-py2.py3-none-any.whl", hash = "sha256:9f47eda37229f68eee03b24b9748937c7dc3868f906e8ba69fbcbdd3bc5dc3e2"},
    {file = "colorama-0.4.4.tar.gz", hash = "sha256:5941b2b48a20143d2267e95b1c2a7603ce057ee39fd88e7329b0c292aa16869b"},
]
graphql-core = [
    {file = "graphql-core-3.2.0.tar.gz", hash = "sha256:86e2a0be008bfde19ef78388de8a725a1d942a9190ca431c24a60837973803ce"},
    {file = "graphql_core-3.2.0-py3-none-any.whl", hash = "sha256:0dda7e63676f119bb3d814621190fedad72fda07a8e9ab780bedd9f1957c6dc6"},
]
mypy = [
    {file = "mypy-0.941-cp310-cp310-macosx_10_9_universal2.whl", hash = "sha256:98f61aad0bb54f797b17da5b82f419e6ce214de0aa7e92211ebee9e40eb04276"},
    {file = "mypy-0.941-cp310-cp310-macosx_10_9_x86_64.whl", hash = "sha256:6a8e1f63357851444940351e98fb3252956a15f2cabe3d698316d7a2d1f1f896"},
    {file = "mypy-0.941-cp310-cp310-macosx_11_0_arm64.whl", hash = "sha256:b30d29251dff4c59b2e5a1fa1bab91ff3e117b4658cb90f76d97702b7a2ae699"},
    {file = "mypy-0.941-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl", hash = "sha256:8eaf55fdf99242a1c8c792247c455565447353914023878beadb79600aac4a2a"},
    {file = "mypy-0.941-cp310-cp310-win_amd64.whl", hash = "sha256:080097eee5393fd740f32c63f9343580aaa0fb1cda0128fd859dfcf081321c3d"},
    {file = "mypy-0.941-cp36-cp36m-macosx_10_9_x86_64.whl", hash = "sha256:f79137d012ff3227866222049af534f25354c07a0d6b9a171dba9f1d6a1fdef4"},
    {file = "mypy-0.941-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl", hash = "sha256:8e5974583a77d630a5868eee18f85ac3093caf76e018c510aeb802b9973304ce"},
    {file = "mypy-0.941-cp36-cp36m-win_amd64.whl", hash = "sha256:0dd441fbacf48e19dc0c5c42fafa72b8e1a0ba0a39309c1af9c84b9397d9b15a"},
    {file = "mypy-0.941-cp37-cp37m-macosx_10_9_x86_64.whl", hash = "sha256:0d3bcbe146247997e03bf030122000998b076b3ac6925b0b6563f46d1ce39b50"},
    {file = "mypy-0.941-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl", hash = "sha256:3bada0cf7b6965627954b3a128903a87cac79a79ccd83b6104912e723ef16c7b"},
    {file = "mypy-0.941-cp37-cp37m-win_amd64.whl", hash = "sha256:eea10982b798ff0ccc3b9e7e42628f932f552c5845066970e67cd6858655d52c"},
    {file = "mypy-0.941-cp38-cp38-macosx_10_9_universal2.whl", hash = "sha256:108f3c7e14a038cf097d2444fa0155462362c6316e3ecb2d70f6dd99cd36084d"},
    {file = "mypy-0.941-cp38-cp38-macosx_10_9_x86_64.whl", hash = "sha256:d61b73c01fc1de799226963f2639af831307fe1556b04b7c25e2b6c267a3bc76"},
    {file = "mypy-0.941-cp38-cp38-macosx_11_0_arm64.whl", hash = "sha256:42c216a33d2bdba08098acaf5bae65b0c8196afeb535ef4b870919a788a27259"},
    {file = "mypy-0.941-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl", hash = "sha256:fc5ecff5a3bbfbe20091b1cad82815507f5ae9c380a3a9bf40f740c70ce30a9b"},
    {file = "mypy-0.941-cp38-cp38-win_amd64.whl", hash = "sha256:bf446223b2e0e4f0a4792938e8d885e8a896834aded5f51be5c3c69566495540"},
    {file = "mypy-0.941-cp39-cp39-macosx_10_9_universal2.whl", hash = "sha256:745071762f32f65e77de6df699366d707fad6c132a660d1342077cbf671ef589"},
    {file = "mypy-0.941-cp39-cp39-macosx_10_9_x86_64.whl", hash = "sha256:465a6ce9ca6268cadfbc27a2a94ddf0412568a6b27640ced229270be4f5d394d"},
    {file = "mypy-0.941-cp39-cp39-macosx_11_0_arm64.whl", hash = "sha256:d051ce0946521eba48e19b25f27f98e5ce4dbc91fff296de76240c46b4464df0"},
    {file = "mypy-0.941-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl", hash = "sha256:818cfc51c25a5dbfd0705f3ac1919fff6971eb0c02e6f1a1f6a017a42405a7c0"},
    {file = "mypy-0.941-cp39-cp39-win_amd64.whl", hash = "sha256:b2ce2788df0c066c2ff4ba7190fa84f18937527c477247e926abeb9b1168b8cc"},
    {file = "mypy-0.941-py3-none-any.whl", hash = "sha256:3cf77f138efb31727ee7197bc824c9d6d7039204ed96756cc0f9ca7d8e8fc2a4"},
    {file = "mypy-0.941.tar.gz", hash = "sha256:cbcc691d8b507d54cb2b8521f0a2a3d4daa477f62fe77f0abba41e5febb377b7"},
]
mypy-extensions = [
    {file = "mypy_extensions-0.4.3-py2.py3-none-any.whl", hash = "sha256:090fedd75945a69ae91ce1303b5824f428daf5a028d2f6ab8a299250a846f15d"},
    {file = "mypy_extensions-0.4.3.tar.gz", hash = "sha256:2d82818f5bb3e369420cb3c4060a7970edba416647068eb4c5343488a6c604a8"},
]
pygments = [
    {file = "Pygments-2.11.2-py3-none-any.whl", hash = "sha256:44238f1b60a76d78fc8ca0528ee429702aae011c265fe6a8dd8b63049ae41c65"},
    {file = "Pygments-2.11.2.tar.gz", hash = "sha256:4e426f72023d88d03b2fa258de560726ce890ff3b630f88c21cbb8b2503b8c6a"},
]
python-dateutil = [
    {file = "python-dateutil-2.8.2.tar.gz", hash = "sha256:0123cacc1627ae19ddf3c27a5de5bd67ee4586fbdd6440d9748f8abb483d3e86"},
    {file = "python_dateutil-2.8.2-py2.py3-none-any.whl", hash = "sha256:961d03dc3453ebbc59dbdea9e4e11c5651520a876d0f4db161e8674aae935da9"},
]
python-multipart = [
    {file = "python-multipart-0.0.5.tar.gz", hash = "sha256:f7bb5f611fc600d15fa47b3974c8aa16e93724513b49b5f95c81e6624c83fa43"},
]
six = [
    {file = "six-1.16.0-py2.py3-none-any.whl", hash = "sha256:8abb2f1d86890a2dfb989f9a77cfcfd3e47c2a354b01111771326f8aa26e0254"},
    {file = "six-1.16.0.tar.gz", hash = "sha256:1e61c37477a1626458e36f7b1d82aa5c9b094fa4802892072e49de9c60c4c926"},
]
strawberry-graphql = [
    {file = "strawberry-graphql-0.103.1.tar.gz", hash = "sha256:296ea0d777200e12114c4897a9fa6573209a9a00cd4b202cf04a6fd48b17f0c8"},
    {file = "strawberry_graphql-0.103.1-py3-none-any.whl", hash = "sha256:f894464f765af4f38a5c4579fdd1874d2c824b067f5f673fb1e45c1f433af648"},
]
tomli = [
    {file = "tomli-2.0.1-py3-none-any.whl", hash = "sha256:939de3e7a6161af0c887ef91b7d41a53e7c5a1ca976325f429cb46ea9bc30ecc"},
    {file = "tomli-2.0.1.tar.gz", hash = "sha256:de526c12914f0c550d15924c62d72abc48d6fe7364aa87328337a31007fe8a4f"},
]
typing-extensions = [
    {file = "typing_extensions-4.1.1-py3-none-any.whl", hash = "sha256:21c85e0fe4b9a155d0799430b0ad741cdce7e359660ccbd8b530613e8df88ce2"},
    {file = "typing_extensions-4.1.1.tar.gz", hash = "sha256:1a9462dcc3347a79b1f1c0271fbe79e844580bb598bafa1ed208b94da3cdcd42"},
]
```

</details>

`strawberry_testing/__init__.py`
```python
from strawberry.dataloader import DataLoader
from typing import Union, TYPE_CHECKING
import asyncio
import dataclasses

@dataclasses.dataclass
class MyRealType:
    id: int

async def load_fn(ids: list[int]) -> list[Union[MyRealType, ValueError]]: # -> list[MyRealType] is a type error
    return [
        MyRealType(id=id) if id > 0 else ValueError("not found")
        for id in ids
    ]

x_loader = DataLoader(load_fn=load_fn)

async def do_it() -> None:
    value = await x_loader.load(1)
    if TYPE_CHECKING:
        reveal_type(value) # Type note: Revealed type is "Union[builtins.str, builtins.ValueError]"

    # assert isinstance(value, MyRealType) # resolves type error
    whatever = value.id # Type error: Item "ValueError" of "Union[MyRealType, ValueError]" has no attribute "id"

    await x_loader.load(-1) # Raises: ValueError: not found

asyncio.run(do_it())
```

mypy:

```
$ poetry run mypy .                                 
strawberry_testing/__init__.py:21: note: Revealed type is "Union[strawberry_testing.MyRealType, builtins.ValueError]"
strawberry_testing/__init__.py:24: error: Item "ValueError" of "Union[MyRealType, ValueError]" has no attribute "id"
Found 1 error in 1 file (checked 1 source file)
```

Runtime:
```
$ poetry run python ./strawberry_testing/__init__.py
Traceback (most recent call last):
  File "./strawberry_testing/__init__.py", line 28, in <module>
    asyncio.run(do_it())
  File ".../lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File ".../lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "./strawberry_testing/__init__.py", line 26, in do_it
    await x_loader.load(-1) # Raises: ValueError: not found
ValueError: not found

```

**Comment by patrick91** [comment]
@huonw ah interesting 😊 I didn't think about that, sounds like a good idea :)

Are you willing to make a pull request? 😊

**Comment by huonw** [comment]
Cool, thanks for validating. I'll give it a go 😄

**Comment by huonw** [comment]
This is fixed by #1737, I just forgot to use the "closes #1736" syntax in the PR for GitHub to close this automatically.
